paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,497 @@
+                                        dst_desc, output, ctx);
+    return *(this->fc_);
+  }
+
+  virtual mkldnn::memory Reorder(const memory::desc& src_desc,
+                                 const memory::desc& dst_desc,
+                                 const void* src_data,
+                                 const std::vector<float>& scale = {1.0f}) {
+    auto mask = scale.size() > 1 ? 1 : 0;
+    mkldnn::primitive_attr attr;
+    attr.set_output_scales(mask, scale);
+    auto src_mem =
+        memory({src_desc, this->engine_}, const_cast<void*>(src_data));
+    auto dst_mem = memory({dst_desc, this->engine_});
+    auto reorder_pd = mkldnn::reorder::primitive_desc(
+        src_mem.get_primitive_desc(), dst_mem.get_primitive_desc(), attr);
+
+    auto reorder = mkldnn::reorder(reorder_pd, src_mem, dst_mem);
+    stream(stream::kind::eager).submit({reorder}).wait();
+
+    return dst_mem;
+  }
+
+  mkldnn::memory TransposeWeights(const Tensor* weights,
+                                  mkldnn::memory::data_type type) {
+    auto dims = framework::vectorize2int(weights->dims());
+    std::swap(dims[0], dims[1]);  // Correct output dimensions
+    auto src_desc = this->CreateMemDescriptor(dims, type, memory::format::io);
+    auto dst_desc = this->CreateMemDescriptor(dims, type, memory::format::oi);
+    // MOD: return Reorder(src_desc, dst_desc, weights->data<T>());
+    return Reorder(src_desc, dst_desc, weights->data<K>());
+  }
+
+  mkldnn::memory UpdateWeights(mkldnn::memory weights,
+                               const std::vector<float>& scale_weight) {
+    auto dims = weights.get_primitive_desc().desc().data.dims;
+    auto ndims = weights.get_primitive_desc().desc().data.ndims;
+    auto weights_dims = std::vector<int>(dims, dims + ndims);
+    auto weights_dt = platform::MKLDNNGetDataType<K>();
+    auto weights_desc = this->CreateMemDescriptor(
+        weights_dims, memory::data_type::s8, memory::format::oi);
+    auto user_weights_desc =
+        this->CreateMemDescriptor(weights_dims, weights_dt, memory::format::oi);
+    return Reorder(user_weights_desc, weights_desc, weights.get_data_handle(),
+                   scale_weight);
+  }
+
+  virtual inner_product_forward CreateFcPrimitive(const memory& src_memory,
+                                                  const memory& weights_memory,
+                                                  const memory::desc& dst_desc,
+                                                  Tensor* output,
+                                                  const ExecutionContext& ctx) {
+    const auto weights_desc = weights_memory.get_primitive_desc().desc();
+    const auto src_desc = src_memory.get_primitive_desc().desc();
+
+    mkldnn::primitive_attr fc_attr;
+
+    bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+    auto scale_weights_data = ctx.Attr<std::vector<float>>("scale_y");
+    auto scale_in_data = ctx.Attr<float>("scale_x");
```

Comment:
Why does y input use a multichannel quantization while x only uses a single scale?

Reply:
I think that's how multichannel quantization designed.

Reply:
done

Reply:
Since as you told me mul is just an fc without bias, it makes sense to have multichannel quantization. Please discard my initial comment.